synapse/tests/test_link.py

```diff
@@ -77,6 +86,154 @@ def test_link_fromurl(self):
         self.eq(link[1].get('user'), 'visi')
         self.eq(link[1].get('passwd'), 'secret')
 
+    def run_nonblocking_pair_recv_tests(self, s1, s2):
+        tenmegs = 10000000
+        s1.setblocking(False)
+        s2.setblocking(False)
+
+        # call recv on a nonblocking socket w/ no data, should return None
+        data = s2.recv(tenmegs)
+        self.none(data)
+
+        # call recv on a closed nonblocking socket w/ no data, should return b''
+        s2.close()
+        data = s2.recv(tenmegs)
+        self.eq(data, b'')
+
+        # call recv w/ negative buffer size on a closed nonblocking socket w/ no data, should return b''
+        data = s2.recv(-1 * tenmegs)
+        self.eq(data, b'')
+
+        s1.close()
+        s2.close() # close already closed socket
+
+    def run_nonblocking_pair_recv_err_tests(self, s1, s2):
+        tenmegs = 10000000
+        s1.setblocking(False)
+        s2.setblocking(False)
+
+        # call recv w/ negative buffer size, resulting in it being finid
+        data = s2.recv(-1 * tenmegs)
+        self.eq(data, b'')
+        self.raises(BrokenPipeError, s1.send, b'wat')
+
+        s1.close()
+        s2.close()
+
+    def run_nonblocking_pair_send_tests(self, s1, s2):
+        tenmegs = 10000000
+        data = os.urandom(tenmegs)
+        s1.setblocking(False)
+        s2.setblocking(False)
+
+        # try to send ten megs, assert that fewer than ten megs are sent
+        send0 = s1.send(data)
```

Comment:
When we call send with a piece of data of this size, the tcp socket will send some of it and return how many bytes were sent. The SSL socket will raise the SSLWantWriteError exception, which we are catching and returning 0 from. This exception can actually be raised after bytes are sent though, so it would return 0 incorrectly. Thoughts?

Reply:
Note that in both the TCP and SSL implementations, we try to call send and then return 0 if we get some exception stating that we tried to send, but the send would block.